Making ELF files compatible

FredCoast opened the pull request:
Creating the functionality for using ELF files, integrating them into the original application

FredCoast committed:
Making ELF files compatible

williballenthin reviewed (commented):
this is such a great start! i'm really excited by the functionality that you already added. 

i think the first order of business is to add tests for some example ELF files. would collect one or two representative ELF files (maybe one shared object and one executable) and add tests for each of the new routines (as reasonable)? you can use the tests for the PE loader as an example. 

thoughts?

FredCoast committed:
Adding some linux test files

nop_elf - /bin/true - equivalent of a nop.exe
ls - /bin/ls - equivalent of K32

Both gotten from a x64 linux mint distro

FredCoast committed:
Added ELF testing, updated the tests to use the current unicorn version, and added tests (replaced ls with libc.so.6)

FredCoast commented:
I've updated it to add tests, sorry it took me so long I was moving into my new university so I've been a bit busy

I have tried to keep the tests used as similar as the ones that you originally used, I mainly just got them from linux mint if that is alright?

Again sorry it took so long

williballenthin commented:
looking good!

the loader tests look reasonable. 

next it might make sense to add some tests around code discovery. like take known functions from IDA/Ghidra and see if lancelot has found them (and triage if not). it's ok to add a test case for every function we expect to find, and even those we know don't work yet, and slowly we can increase our test passing rate. it might take quite a bit of time to get the ELF code analysis working well - i'm not sure yet

FredCoast commented:
Sounds good, so to double check create more functionality and tests like the ones in:
- export.rs
- entrypoints.rs
- control_flow_guard.rs
- runtime_functions.rs
- noret_imports.rs
(but relative to elf files)
until the results are similar to those in IDA/Ghidra


I don't mind working on it over the long run, I'm learning lots already so its been very useful, thanks!

williballenthin commented:
Control Flow Guard and Runtime Functions are Windows-centric structures, so they won't be relevant for ELF files. but perhaps there are other Linux or generic ELF structures we could use too. 

it's fine to start with exactly whatever you need for QS. otherwise you risk getting distracted. but as long as you're having fun and learning, it doesn't matter! 

i'd probably start with entry points and exports, and then do an hour or two of investigation of what other techniques there are to find functions. that should be a solid start.

FredCoast commented:
I was just using them as an example, like how the tests might look/where they are kept,

I'm currently working on implementing undefined thunks through the procedure linkage table, as I realised that I was hardly picking up any functions and this seems to be where most of that fall off is (at least based on the tests I've been using).

And with QS I would rather wait for Moritz to reply before doing too much in case I spend time on something that isn't relevant by accident

FredCoast committed:
Adding expanded PLT function detection, and removing a no longer used test

FredCoast committed:
Collecting plt.got and entries

FredCoast committed:
Adding tests and seperating out files

FredCoast commented:
I believe I've added PLT, entry points and exports but I'm struggling to get it to pick up some of the functions that Ghidra has identified, I've tried to research potential methods for function detection with most resources saying heuristics and cfg being the main methods. 

So I have attempted to use your implementations for those methods, as I am assuming that as we are working with x86-64 the methods would work for both. However I don't seem to be finding the functions so I was wondering if you had any pointers?

Also should the exports pick up global symbols and functions or just functions as Ghidra groups both from what I can tell?

FredCoast commented:
Yeah I think that is the use case, 

that would be pretty cool, sort of like instead of showing strings by sections instead show them by function?

I'm happy for it to be merged now, but at the end of the day its your project so I'm happy with whatever you decide

FredCoast committed:
Adding FDE analysis

FredCoast commented:
Done! 

your is_probably_code was a great solution, so thanks for that! There are still a couple of items which I can't figure out how to filter because Ghidra says nothing whatsoever about them.

I also had to use is_probably_code on the CFG analysis as adding the FDE caused it to pick up loads of subroutines. Which were some of the bad instructions, so I'm just double checking if that is alright

williballenthin commented:
> Are u saying to keep the output consistent by filtering out the thunks?

we should keep them around, and even mark them as code and functions (like we do in PE), even if they appear to be trivial (perhaps just `jmp $elsewhere`).

> And should I not show the imported functions in general?

again, i think we should try to capture the fact we have these references to imports (either as thunks or GOT entries or whatever). if the existing thunks infrastructure isn't enough, at least as PE uses it, feel free to introduce a new concept that works better for ELF (i'm wondering if GOT entries will work or not, i'm not sure).

FredCoast commented:
Ill create a new function to see if there is overlap between the plt.got/got and something returned by the CFG thunk function, then store it as an import which I believe should solve this issue.

FredCoast committed:
Filtering out plt entries from functions and storing them as imports

FredCoast commented:
Filtering out the GOT worked, I've also noticed that many functions get called through 

LEA RDI, [*function*]

I believe that there is some implementation for it, however I don't think they are being picked up properly, would it be alright if I modified the implementation to try to get it to work?

FredCoast commented:
Not related to this but you were just referenced in one of my lectures, and its readings!

FredCoast committed:
Patterns implementation, is picking up some functions which match the fde, and some which don't (in libc)

williballenthin commented:
nice!

many of the patterns you added are fairly specific to MSVC. would you consider taking a look at the other patterns that ghidra has for GCC and seeing how well they work for ELF files?

https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/x86/data/patterns/x86-64gcc_patterns.xml

for example

FredCoast committed:
Adding patterns from Ghidra

FredCoast committed:
Updating the pre-patterns and un-commenting the fde

FredCoast commented:
Ah makes sense, I'll leave them uncommented for now as admittedly I don't know. I've updated it for the gcc patterns, it's resulted in many more being picked up so thanks for the suggestion!

Out of curiosity why shouldn't personal experience be used for something like this?

FredCoast commented:
Ah so the implementations used now are subject to the bias of current trends and usage, however this can become quickly outdated when being considered in the future, and facts make long term development more sustainable.

Thanks for explaining it

FredCoast committed:
Adding symtab

FredCoast committed:
Added DWARF capabilities and added a test specifically for it

FredCoast commented:
I've added symtab, dynsym and DWARF detection, 

I had some reservations about adding them in general as they will most likely just get stripped from the file but oh well

Ill have a look at ghidra and see what else they have that I can implement soon

williballenthin commented:
nice! i found a copy as well and will read it tonight

williballenthin commented:
do you mind sharing the document as an upload here? i don't have access to the o365 drive. edit: maybe that drive has the dataset itself. if it's important i can request access, let me know. 

i did the nucleus algorithm early on, and it did find good code, though i recall not having good handling for jump tables was causing trouble. so i "went back to basics" and used more conservative algorithms. now's probably a great time to reassess and perhaps try again (and document our lessons learned explicitly!).

FredCoast commented:
It was like 18gb of test files, unfortunately I have also lost access it seems, nothing that I am too fussed about to be honest.

Cool I will do the traceback and then nucleus (with explicit documentation)

FredCoast committed:
Adding call targets

FredCoast committed:
Modifying to work with the tests

williballenthin commented:
take your time :-) work on this when you want to and when it brings you joy, no expectations or pressure from me

FredCoast committed:
Adding unconditional jump targets, working on reducing false positives now

FredCoast committed:
Binned off JMP, created too many false positives, made it only check executable code

FredCoast committed:
Fixing FDE table

FredCoast commented:
I'm so sorry I am probably going to have to stop working on this for a while (at least until May),

I am starting an internship, and I don't think I will have the time to fit this in.

I've really enjoyed working on this, learning about Rust and about development in general. 

And thank you for the advice that you have given me over the past couple of months, I hope I find time to keep working on this and sorry to leave it in such an unfinished state.

FredCoast commented:
Hello again, I wanted to run something by you

Could looking for parameters in disassembled code be a sign that a function boundary is nearby?

As call symbols are so common, if there were common parameter heuristics before them would it potentially be a sign that there is a function start instead of an if statement for example?
